Ft codefirstdb - MongoDb Setup + Initial Models

townerr opened the pull request:
Added DataAccess Project:
   - Setup initial MongoDB model classes
   - Setup db settings classes used in startup to config db
   - Added MongoDB nuget packages
   
Added MongoDB config/service to the startup.cs

The db settings are located in user secrets.

townerr committed:
Initial mongo collection models and settings class
Added mongo nuget packages

townerr committed:
Add Mongo startup confid
Add mongo settings to user secrets
Added notification collection to user

townerr committed:
Removed total unread notifications

townerr commented:
Removed total unread notifications since that can be calculated on the frontend

mendac017 reviewed (approved):
This looks good. You let me know you tested it on your local database.

escalonn commented on FakebookNotifications/FakebookNotifications.WebApi/Startup.cs:
```diff
@@ -26,6 +29,17 @@ public Startup(IConfiguration configuration)
+
+            //Mongo add db service
+            services.AddSingleton<INotificationsDatabaseSettings>(sp =>
+                sp.GetRequiredService<IOptions<NotificationsDatabaseSettings>>().Value);
```
i'd add package reference Microsoft.Extensions.Options to the data access layer and then just request IOptions<NotificationsDatabaseSettings> where you wind up needing it, rather than reregister the service. IOptions is a fine dependency to take, it's pretty generic.

drop the commented out code below

escalonn commented on FakebookNotifications/FakebookNotifications.DataAccess/Models/Notification.cs:
```diff
@@ -0,0 +1,33 @@
+        [BsonElement("user")]
+        public User LoggedInUser { get; set; }
+
+        //user that triggered a notifiaction by commenting on a user post or tagging a user etc.
```
spelling

escalonn commented on FakebookNotifications/FakebookNotifications.DataAccess/Models/Notification.cs:
```diff
@@ -0,0 +1,33 @@
+
+        //user logged in to receive notifications
+        [BsonElement("user")]
+        public User LoggedInUser { get; set; }
```
don't put circular references in your mongodb object model without a pressing need. this could be just a user ID.
same with the triggeruser below.

escalonn commented on FakebookNotifications/FakebookNotifications.DataAccess/Models/Interfaces/INotificationsDatabaseSettings.cs:
```diff
@@ -0,0 +1,13 @@
+
+        //DB Settings
+        string ConnectionString { get; }
+        string DatabaseName { get; }
```
the non-secret values of this configuration should just go in appsettings.json. less for us to keep track of with user secrets + env vars

escalonn commented on FakebookNotifications/FakebookNotifications.DataAccess/Models/Notification.cs:
```diff
@@ -0,0 +1,33 @@
+
+        //notification type: post/comment/tag
+        [BsonElement("type")]
+        public string Type { get; set; }
```
surely we need some kind of "target" or "reason" id here too - post id if it's a post, comment id if it's a comment, etc.

escalonn commented on FakebookNotifications/FakebookNotifications.DataAccess/Models/Notification.cs:
```diff
@@ -0,0 +1,33 @@
+
+        //date of notification
+        [BsonElement("date")]
+        public BsonDateTime Date { get; set; }
```
keep in mind we want to store this data such that it's efficient to do things like "get the most recent 20 notifications for user X", so we probably want an index on Date. figure out how to configure that when the time comes

townerr committed:
Removed DB settings from being a service in startup.
Added Notification Repo and Interface.
Added db settings to repo.
Change users in notificaion model to be email ids.
Moved non sensitive stuff from user secret to appsettings.

townerr committed:
Removed all of db settings service config

townerr commented:
Removed the User types on notification fields and replaced them with strings to be user ids.
Fixed spelling on comment.
Made type a KeyValuePair to have a string type and int id of what made the notification

townerr committed:
Added a NotificationsRepo for settings and collections
Removed settings stuff from Repo and replaced witr context.

townerr commented:
Removed the db settings service stuff from the startup.cs and created a NotificationsRepo and NotificationsContext to handle the db settings instead.

The NotificationsContext handles creating the db client and database and methods for creating the collections.
The NotificationsRepo just creates the context object that will be used in the repo.

escalonn commented:
/azp run

escalonn commented:
fix issues identified by the pipeline (updating the dockerfile, for this first error)
for some reason the pipeline didn't run until i triggered it myself there; it should run on subsequent pushes though?

escalonn commented on FakebookNotifications/FakebookNotifications.DataAccess/NotificationsRepo.cs:
```diff
@@ -0,0 +1,24 @@
+        public NotificationsRepo(IOptions<NotificationsDatabaseSettings> settings)
+        {
+            //create context from settings
+            _context = new NotificationsContext(settings);
```
there's some "new is glue" issues here, decouple to allow unit testing. (eshoponweb opted for functional testing only on that locations.api) but really we can't start implementing the repository until we've defined the domain model, so this class just shouldn't be here yet.

escalonn commented on FakebookNotifications/FakebookNotifications.Testing/FakebookNotifications.Testing.csproj:
```diff
@@ -8,6 +8,8 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.7.1" />
+    <PackageReference Include="mongocsharpdriver" Version="2.11.5" />
```
this package is deprecated

townerr committed:
Merge branch 'us-notifications' into ft-codefirstdb

escalonn commented on FakebookNotifications/FakebookNotifications.WebApi/Startup.cs:
```diff
@@ -26,7 +27,6 @@ public Startup(IConfiguration configuration)
         // This method gets called by the runtime. Use this method to add services to the container.
         public void ConfigureServices(IServiceCollection services)
         {
-
```
the two diffs to this file are unnecessary

townerr committed:
Mongo Repo Tests

mattmarnien committed:
updated dockerfile to take dataccessprj into account

townerr committed:
add some seed data json files for mongodb

SukafuDesu committed:
[Enhancement] Added Skeleton for FakebookNotifications (#11)

* [Enhancement] Added Skeleton for FakebookNotifications

* Added Interfaces
* Added Domain Class for User and Notifications

* [Enhancement] Updated Repos According to Docs
* https://docs.google.com/document/d/1AuPOR2f3VtaDPO6iRbpaVjzjzYFbjV3-m-VCY6U3vGA/edit
* Updated NotificationsRepo to implement interface
* Create/Updated UserRepo to implement interface

* Made Repos public

SukafuDesu committed:
[Updated Dockerfile]

* Dockerfile should now build with new domains library

townerr committed:
Changed Repo to get context instead of creating a new one.
Configured db settings to get from appsettings.
Added repo and context to be servivces for dependency injection.
Removed deprecated package.

escalonn commented on FakebookNotifications/Dockerfile:
```diff
@@ -6,12 +6,16 @@ WORKDIR /app
 COPY FakebookNotifications.WebApi/. ./FakebookNotifications.WebApi/
 COPY FakebookNotifications.Testing/. ./FakebookNotifications.Testing/
+COPY FakebookNotifications.DataAccess/. ./FakebookNotifications.DataAccess/
+COPY FakebookNotifications.Domains/. ./FakebookNotifications.Domains/
```
these four lines (15-18) are actually pointless, they should be removed

escalonn commented on FakebookNotifications/FakebookNotifications.Domains/Interfaces/INotificationsRepo.cs:
```diff
@@ -0,0 +1,33 @@
+        /// <summary>
+        ///  Add a Notification to the databse
+        /// </summary>
+        /// <returns> A bool that let's us know if we passed or failed</returns>
```
documentation is clumsy. fix typos, document all the return values and params.
https://www.mking.net/blog/documenting-task-returns

escalonn reviewed (changes requested):
three types of PRs to us-* branches:
- focused code that you want to not unit test (maybe we'll agree).
- unit tests and the minimum code needed to make them compile.
- the minimum code needed to make some failing unit tests pass.

this PR needs to become one and only one of those types, with the rest of the code in some different PR.

townerr committed:
Removed domain code from branch

townerr committed:
remove domain stuff

townerr committed:
fixed docker file

townerr committed:
fix docker error

townerr commented:
All the domain skeleton stuff that was added in has been moved to its own branch, and the documentation should be fixed there.
Fixed the docker file.

mendac017 commented on FakebookNotifications/FakebookNotifications.Testing/MongoDbTests.cs:
```diff
@@ -0,0 +1,40 @@
+        [Fact]
+        public void MongoDbConnectionTest()
+        {
+            //Arrange
```
remove these comments, or make them fail as not implemented.

townerr committed:
Added mongo tests

Commit:
Merge branch 'us-notifications' into ft-codefirstdb

escalonn commented on FakebookNotifications/FakebookNotifications.DataAccess/NotificationsContext.cs:
```diff
@@ -0,0 +1,61 @@
+                Debug.WriteLine("Database Connection Successfull.");
+                return true;
+            }
+            else
```
the result of a constructor will never be null. maybe you mean to catch some specific exception here. yeah i see that code i showed you does this, lol.

escalonn commented on FakebookNotifications/FakebookNotifications.DataAccess/NotificationsContext.cs:
```diff
@@ -0,0 +1,61 @@
+        {
+            get
+            {
+                return _database.GetCollection<User>(_settings.UserCollection);
```
this setup isn't so good because _database is null before you call Connect().

escalonn commented on FakebookNotifications/FakebookNotifications.DataAccess/NotificationsContext.cs:
```diff
@@ -0,0 +1,61 @@
+
+namespace FakebookNotifications.DataAccess
+{
+    public class NotificationsContext
```
the class should go through TDD, maybe not every part of it can be unit tested but some parts can, so this is the wrong kind of PR to add it in.

escalonn commented on FakebookNotifications/FakebookNotifications.DataAccess/NotificationsRepo.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace FakebookNotifications.DataAccess
+{
+    public class NotificationsRepo: INotificationsRepo
```
should use dependency inversion etc., but these classes should not exist yet. you could make their interfaces, but not the classes yet.

escalonn commented on FakebookNotifications/FakebookNotifications.Testing/MongoDbTests.cs:
```diff
@@ -0,0 +1,56 @@
+    {
+        //Test if connection to database is successfull
+        [Fact]
+        public void MongoDbConnectionTest()
```
these tests belong in a separate tdd-fail PR.

escalonn commented on FakebookNotifications/FakebookNotifications.Testing/MongoDbTests.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            //Arrange
+            var mockSettings = Mock.Of<IOptions<NotificationsDatabaseSettings>>();
+            var mockContext = new Mock<NotificationsContext>(mockSettings.Value);
```
this is misusing Mock. you have to have a very precise idea of what this test is testing. if it's the context's Connect method, then mocking the context makes zero sense. you mock the _dependencies_ of the test subject. this still does test the code, because mocking a class (as opposed to an interface) is using the real implementations by default, but it's semantically wrong.

escalonn commented:
get up to date with upstream branch and follow our new PR policies.

townerr committed:
Removed tests to be added to a test branch

townerr committed:
Merge branch 'ft-codefirstdb' of https://github.com/2011-fakebook-project3/notifications into ft-codefirstdb
